src/common/Cryptography/CryptoHash.h

```diff
@@ -0,0 +1,111 @@
+#include "Define.h"
+#include "Errors.h"
+#include <array>
+#include <string>
+#include <string_view>
+#include <openssl/evp.h>
+
+class BigNumber;
+
+namespace Trinity
+{
+    namespace Impl
+    {
+        struct GenericHashImpl
+        {
+            typedef EVP_MD const* (*HashCreator)();
+
+#if defined(OPENSSL_VERSION_NUMBER) && OPENSSL_VERSION_NUMBER < 0x10100000L
+            static EVP_MD_CTX* MakeCTX() { return EVP_MD_CTX_create(); }
+            void DestroyCTX(EVP_MD_CTX* ctx) { EVP_MD_CTX_destroy(ctx); }
+#else
+            static EVP_MD_CTX* MakeCTX() { return EVP_MD_CTX_new(); }
+            static void DestroyCTX(EVP_MD_CTX* ctx) { EVP_MD_CTX_free(ctx); }
+#endif
+        };
+
+        template <GenericHashImpl::HashCreator HashCreator, size_t DigestLength>
+        class GenericHash
+        {
+            public:
+                static constexpr size_t DIGEST_LENGTH = DigestLength;
+                using Digest = std::array<uint8, DIGEST_LENGTH>;
+
+                static Digest GetDigestOf(uint8 const* dta, int len)
+                {
+                    GenericHash hash;
+                    hash.UpdateData(dta, len);
+                    hash.Finalize();
+                    return hash.GetDigest();
+                }
+                template <typename... Ts>
+                static auto GetDigestOf(Ts&&... pack) -> std::enable_if_t<!(std::is_integral_v<std::decay_t<Ts>> || ...), Digest>
+                {
+                    GenericHash hash;
+                    (hash.UpdateData(std::forward<Ts>(pack)), ...);
+                    hash.Finalize();
+                    return hash.GetDigest();
+                }
+
+                GenericHash() : _ctx(GenericHashImpl::MakeCTX())
+                {
+                    int result = EVP_DigestInit_ex(_ctx, HashCreator(), nullptr);
+                    ASSERT(result == 1);
+                }
+
+                void UpdateData(uint8 const* dta, size_t len)
+                {
+                    int result = EVP_DigestUpdate(_ctx, dta, len);
+                    ASSERT(result == 1);
+                }
+                void UpdateData(std::string_view str) { UpdateData(reinterpret_cast<uint8 const*>(str.data()), str.size()); }
+                void UpdateData(std::string const& str) { UpdateData(std::string_view(str)); } /* explicit overload to avoid using the container template */
+                void UpdateData(char const* str) { UpdateData(std::string_view(str)); } /* explicit overload to avoid using the container template */
+                template <typename Container> void UpdateData(Container const& c) { UpdateData(std::data(c), std::size(c)); }
+
+                void Finalize()
+                {
+                    uint32 length;
+                    int result = EVP_DigestFinal_ex(_ctx, _digest.data(), &length);
+                    ASSERT(result == 1);
+                    ASSERT(length == DIGEST_LENGTH);
+                    _ctx = nullptr;
```

Comment:
Either call `EVP_MD_CTX_free` or reuse the same ctx (has to first inited with `EVP_DigestInit_ex` again).